[Web:Discover] Report events

kimlisa opened the pull request:
part of https://github.com/gravitational/teleport/pull/18546
depends on https://github.com/gravitational/teleport/pull/20207

note: recommend reviewing by commit, i tried to contain specific changes

#### Description
Report events on the following:
 - status abort: browser refresh, back/forward button, going to different routes from same page, confirming exit from the window prompt
 - status success: user clicks on "next" button to go to the next step
 - status error: any api request failures (eg network failure), user error, timeout errors (polling for resources), unsuccessful connection testing

#### Screenshot of json request examples
INIT (when user goes to the discover screen for first time, note the empty resource)
![image](https://user-images.githubusercontent.com/43280172/214154431-ff57f2a1-7024-4008-a125-da85ff365f2c.png)

emitted when user clicks "next"
![image](https://user-images.githubusercontent.com/43280172/214154515-de6e4fa0-8dc6-46fa-abcd-7defecdf9775.png)

skipped example
![image](https://user-images.githubusercontent.com/43280172/214154574-abacf230-95e8-4e4e-ae5a-80610bdbb913.png)

[Test connection failed example](https://github.com/gravitational/teleport/pull/20581/files#diff-afa484093fef884d90a09a4690b22765894d375e490a4654e2a9f419dcf496b1R54)
![image](https://user-images.githubusercontent.com/43280172/214154692-b346f77f-0c0f-49af-b673-206500cd2a71.png)

Abort example
![image](https://user-images.githubusercontent.com/43280172/214154760-08347f6b-38c0-4153-9679-6abb4340fb1d.png)


#### Before merge
- [x] Backend related event PRs have to be merged first (otherwise, if we go with partial events, logic needs to be placed to prevent unsupported events from being emitted if we don't want the error logs to show on user browser)
- [x] Update `userEvent/types.ts > DiscoverEvent` with actual event names after required PRs have been created

kimlisa committed:
Define event types

kimlisa committed:
Define event names for each flow

kimlisa committed:
Report events upon start, exiting, and next steps

kimlisa committed:
Update json property name change

kimlisa committed:
Emit user skipped event

kimlisa committed:
Emit error events

kimlisa commented on web/packages/teleport/src/Discover/Desktop/index.tsx:
```diff
@@ -46,6 +47,7 @@ export const DesktopResource: Resource = {
   views: [
     {
       title: 'Select Resource',
+      eventName: DiscoverEvent.ResourceSelection,
```
@ryanclark @marcoandredinis i think we need these additional event names (i copied and pasted the file and put a comment marker to see what i'm talking about, since it's hard to see with git diff).

i am unsure what to do with steps `Connect Teleport` and `Discover Desktops`. i think these two steps are combined under `Deploy Service` for other resources  (eg `DownloadScript` or `HelmChart`). could be that desktops works slightly different?

```ts
views: [
    {
      title: 'Select Resource',
      eventName: DiscoverEvent.ResourceSelection,
    },
    {
      title: 'Install Active Directory',
      component: InstallActiveDirectory,  // <------- need event here, user has to manually do stuff here (sim to like database mutualTLS)
    },
    {
      title: 'Connect Teleport',
      component: ConnectTeleport,
      eventName: DiscoverEvent.DeployService,     // <---- wasn't sure of this one, this step you create a join token, and user has to run it?
    },
    {
      title: 'Discover Desktops',
      component: DiscoverDesktops,   // <------- this step is where we "poll" for desktops
    },
    {
      title: 'Finished',
      component: Finished,
      eventName: DiscoverEvent.Completed,
      hide: true,
    },
  ],
```

marcoandredinis replied:
For Kube/HelmChart, I think we could use the DeployService step
Seems equivalent, right?

Sorry, I didn't knew how Discover Desktop works.
I agree that we need more events.
Your mappings look correct.

New event for Install Active Directory :+1: 

Connect Teleport seems to match the DeployService event.
Given that we have 4 sub-steps, the user can leave at any point, which could be useful to understand how easy the guide is.
To know where the user stopped or got an error, we could either create 4 distinct events or add some extra info to the error message in case of `error` or `abort` status.
What do you think?
Pinging @ryanclark as well.
If we end up with 4 steps, IMO, we should use the DeployService anyway as the latest step (so we would only create 3 new event names)

For the Discover Desktops I agree that it should be a new event.
Would it be useful to have some extra info? Like how many resources were discovered?
Maybe on a 2nd iteration

ryanclark replied:
Agreed on adding more steps to cover the missing pages

> If we end up with 4 steps, IMO, we should use the DeployService anyway as the latest step (so we would only create 3 new event names)

Agreed on this, so we can keep the last step consistent across the board

> For the Discover Desktops I agree that it should be a new event.
> Would it be useful to have some extra info? Like how many resources were discovered?
> Maybe on a 2nd iteration

It wouldn't hurt to add this info as we already have it for free.

kimlisa replied:
@marcoandredinis after talking with @ryanclark, we need the following events, we have substeps within steps for desktops:

```ts
views: [
    {
      title: 'Select Resource',
      eventName: DiscoverEvent.ResourceSelection,
    },
    {
      title: 'Install Active Directory',
      component: InstallActiveDirectory,  // <------- need event 
    },
    {
      title: 'Connect Teleport',
      component: ConnectTeleport,
        // This step has substeps worth emitting events:
        // 1) script that configures active directory        <----- need event
        // 2) asking user to copy and paste outputted teleport config     // skipping, as we think it's not useful to emit
        // 3) asking user to create `teleport.yaml` and copy and paste   // skipping
        // 4) deploy the desktop service                     <----- use existing `DiscoverEvent.DeployService` 
    },
    {
      title: 'Discover Desktops',
      component: DiscoverDesktops,          // <------- need event (waiting for desktops to be discovered)
    },
    {
      title: 'Finished',
      component: Finished,
      eventName: DiscoverEvent.Completed,
      hide: true,
    },
  ],
```

so we'll need (you can change the name)
- DiscoverEvent.InstallActiveDirectory
- DiscoverEvent.ConfigureActiveDirectory
- DiscoverEvent.DiscoverDesktops

kimlisa replied:
done wiring here: https://github.com/gravitational/teleport/pull/20581/commits/dd6f2647e83feb7d8fc9e999af6180fb7d849340

marcoandredinis replied:
What do you think about a generic Discovered Resources?
After looking into the new Database Flow, it seems we could re-use it there

kimlisa replied:
done: https://github.com/gravitational/teleport/pull/20581/commits/74a142dda9fdb071c4ea8d9f6345d8f828730e77

kimlisa commented:
@marcoandredinis i tagged you for this part (and the comment about desktop specific event names): https://github.com/gravitational/teleport/pull/20581/files#diff-a26c91eae58f21a210af8ed03e545ecb0eb2d2121727c176c6d4e94357961816

kimlisa committed:
Handle unloading event for 'completed' step

kimlisa commented on web/packages/teleport/src/Discover/useDiscover.tsx:
```diff
@@ -100,27 +126,86 @@ export function DiscoverProvider<T = any>(
 
+  useEffect(() => {
+    const emitAbortOrSuccessEvent = () => {
+      if (ref.current.currEventName === DiscoverEvent.Completed) {
+        emitEvent({ stepStatus: DiscoverEventStatus.Success });
```
i didn't think it would make sense to send a `abort` signal on the completed step (it's just letting user know they finished), but i can make it so that if user `does not click` on the two button we provided (as shown in picture), i can send an abort signal instead

![image](https://user-images.githubusercontent.com/43280172/214180903-e3646b75-ce7c-401c-adef-81f47beb1d89.png)

marcoandredinis replied:
Looks good to me :+1:

kimlisa committed:
Move updating event state into hook

kimlisa committed:
Finish emitting events for desktop

kimlisa committed:
Address CR: rename DiscoverDesktop to DiscoverResources

ryanclark commented on web/packages/teleport/src/Discover/useDiscover.tsx:
```diff
@@ -57,21 +65,38 @@ interface DiscoverContextState<T = any> {
+  const [eventState, setEventState] = useState<EventState>();
+  // Use the latest state for async use.
+  const ref = React.useRef<EventState>();
+  useEffect(() => {
+    ref.current = eventState;
+  }, [eventState]);
```
Is this needed? 

Would having `eventState` as dependencies of the `useEffect`s and change the functions to `useCallback` with `eventState` as a dependency work instead?

kimlisa replied:
you're right, i learned that the cleanup func gets called on each call on the useeffect: https://github.com/gravitational/teleport/pull/20581/commits/55d631d8a8379eaea88996325c6cd1d87773f0b8

kimlisa committed:
Various changes:

- address CR: remove use of ref
- diagnostic traces can also be a success, only emit event on failed traces
- add mysql to discover events
- consistent naming on exit prompt

kimlisa committed:
Update and add new event names

kimlisa committed:
Update tests

marcoandredinis reviewed (commented):
LGTM (I've only reviewed the event metadata information)

kimlisa committed:
Emit events for unimplemented flows (apps and other dbs)

kimlisa committed:
Redo emitting events for test connection

Send event for EACH test connection failure or success.
If a user does not do any testing, send a skipped event.

marcoandredinis reviewed (approved):
Go bits look good to me
Bot approving the remaining

kimlisa committed:
Merge branch 'master' into lisa/web/discover/events
